src/python/WMCore/Services/TagCollector/TagCollector.py

```diff
@@ -60,7 +63,7 @@ def _getResult(self, callname="", clearCache=False,
 
         # Note cFile is just the base name pattern, args 
         # are also considered for the end filename in the method below
-        f = self.refreshCache(cFile, callname, args, encoder=encoder,
+        f = self.refreshCache(cFile, callname, args, encoder=encoder, decoder=decodeBytesToUnicode,
```

Comment:
Do we need this in py2 also? 
Because for py3 we should already have it decoded to unicode by https://github.com/dmwm/WMCore/blob/79868333b62c26252035bf2fb57fd480f6c6ec66/src/python/WMCore/Services/Requests.py#L308-L309 
Do you remember why you added this?

Reply:
It was necessary to fix the `CMSSW_t` unit test. This is the change that made all those WMSpec_t/StdSpecs_t unit tests to work as well.

Hmm, looking at the code you pointed out, it should indeed be doing the same job. I might be wrong, but I believe the code is entering in this condition:
https://github.com/dmwm/WMCore/blob/79868333b62c26252035bf2fb57fd480f6c6ec66/src/python/WMCore/Services/Requests.py#L306

thus simply converting things to string. If I set `decoder=False` on this module, I think it would fix the problem, even though the result is exactly the same than passing an explicit decoder.

Reply:
The more I think about this the more I like your approach and the more I want to remove these lines  https://github.com/dmwm/WMCore/blob/79868333b62c26252035bf2fb57fd480f6c6ec66/src/python/WMCore/Services/Requests.py#L308-L309, so i would stick with your option here.

If I have to guess, when running `CMSSW_t`, your result was `bytes` and `decoder` was set to `True` in some keyword argument, so your strings ended up being unicode strings prepended with a `b'content of the original bytes string'`. In these cases my `if PY3:  result = decodeBytesToUnicode(result)` does not do much, and we should specify the decode that we want case by case.

Let me see if i can why i added those lines and if we can get rid of them!

Reply:
Ok, i added it in #10522, let's see what we can do

Reply:
https://github.com/dmwm/WMCore/blob/79868333b62c26252035bf2fb57fd480f6c6ec66/test/python/WMCore_t/Services_t/Requests_t.py#L348-L353

This is why I added `if PY3: result = decodeBytesToUnicode(result)`: if we make requests with `decoder=False, in py2 we have a bytes string `"html"` and we look for it into the result of `makeRequest`, while in py3 `"html"` is unicode. In py3, we can 
1. convert the result from `makeRequest` to unicode by default even if no decode is set (which means adding `if PY3: result = decodeBytesToUnicode(result)`, as done in #10522 ), or 
2. we can never use `decoder=False`, or 
3. we always decode every time after calling `makeRequest` if necessary (maybe not the smartest option, since we already have the possibility of using the `decoder` argument).

I have no strong preferences, but maybe 2. with

```diff
- out = req.makeRequest('/', decoder=False)
+ out = req.makeRequest('/', decoder=lambda x: decodeBytesToUnicodeConditional(x, condition=PY3))
```

is what make more sense to me right now (i can change mind pretty quickly, though)

Reply:
I think that this may cause some problems. Currently, `self.decode(result)` is returning `result.__str__()`, so if `result` is an object (let's say a dictionary), we get its string representation. This should not happen to us in `Requests.makeRequest` (because for example we are not decoding from strings to objects the responses from `pycurl_manager`), but it's something that may happen in the future.
Moreover, i think that returning unicode in py2 may cause a lot of troubles. 

However, I like your idea of combining the `decoder` argument with `decodeBytesToUnicode`, so i propose

```python
elif decoder is not False:
    result = decodeBytesToUnicodeConditional(result, condition=PY3)
    result = self.decode(result)
```

In this way
- py2, decoder not False, result is bytes: it is not touched and we return bytes (which is `str` and wwe are happy)
- py2, decoder not False, result is object: `result = self.decode(result)` convert it to bytes
- py3, decoder not False, result is bytes: it is converted to unicode, untouched by `result = self.decode(result)`
- py3, decoder not false, result is object: untouched by `decodeBytesToUnicode`, converted to unicode by result = self.decode(result)

We have been seeing a lot of problems in py3 (for example i just noticed that `test/python/WMCore_t/Services_t/Requests_t.py:TestRequests.testSecureWithProxy_with_pycurl` is failing in py3) because when we have `decoder` not False, `result = self.decode(result)` is prepending it with the characters `b''` and returning a unicode string. My new approach should be more robust. I can create a new issue and provide this fix, so we see how it behaves against unittests

Reply:
Just a reminder for the two of us: our last approach (for example the one in this commit https://github.com/dmwm/WMCore/pull/10588/commits/886209371d875a5f2f5f7c3fe5e654d639872fec ) will make all uses of `makeRequest(..., decoder=False)` fail in py3. it seems to me that we are using `decoder=False` only in some unittests, so we should not really be affected by this choice